src/lib.rs

```diff
@@ -2,269 +2,209 @@
-        let (dir_handle, calls) = create_echo_bins(&["cat", "bzcat"]).unwrap();
-        uncompress("test.bz2", &dir_handle.path().join("target"), Kind::BZip2)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.bz2", "bzcat"]);
-    }
+        let mut pipe = Pipe {
+            reader: Box::new(r),
+            buffer: vec![0; BUFFER_SIZE],
+        };
+
+        from_archive(
+            archive_read_open(
+                source,
+                (&mut pipe as *mut Pipe) as *mut c_void,
+                None,
+                Some(stream_read_callback),
+                None,
+            ),
+            source,
+        )?;
 
-    #[test]
-    fn uncompress_xz() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "xzcat"]).unwrap();
-        uncompress("test.xz", &dir_handle.path().join("target"), Kind::Xz)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.xz", "xzcat"]);
-    }
+        f(source, entry)?;
 
-    #[test]
-    fn uncompress_lzma() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "lzcat"]).unwrap();
-        uncompress("test.lzma", &dir_handle.path().join("target"), Kind::LZMA)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.lzma", "lzcat"]);
+        from_archive(archive_read_close(source), source)?;
+        from_archive(archive_read_free(source), source)?;
+        archive_entry_free(entry);
     }
+    Ok(())
+}
 
-    #[test]
-    fn uncompress_lzip() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "lzip"]).unwrap();
-        uncompress("test.lzip", &dir_handle.path().join("target"), Kind::LZip)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.lzip", "lzip -dc"]);
+unsafe fn read_data_block<W>(mut target: W, ar: *mut archive) -> Result<()>
+where
+    W: std::io::Write,
+{
+    let mut offset = 0;
+    let mut size = 0;
+    let mut buffer = std::ptr::null();
+
+    loop {
+        match archive_read_data_block(ar, &mut buffer, &mut size, &mut offset) {
+            ARCHIVE_EOF => return Ok(()),
+            ARCHIVE_OK => {
+                let content = slice::from_raw_parts(buffer as *const u8, size);
+                target.write_all(content)?;
+            }
+            _ => return Err(Error::from(ar)),
+        }
     }
+}
 
-    #[test]
-    fn uncompress_zip() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "unzip"]).unwrap();
-        uncompress("test.zip", &dir_handle.path().join("target"), Kind::Zip)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.zip", "unzip - -o -d"]);
+unsafe extern "C" fn stream_read_callback(
+    handle: *mut archive,
+    data: *mut c_void,
+    buff: *mut *const c_void,
+) -> la_ssize_t {
+    let pipe = (data as *mut Pipe).as_mut().unwrap();
+    *buff = pipe.buffer.as_mut_ptr() as *mut c_void;
```

Comment:
Mutability here is actually not necessary it seams. Try just: `*buff = pipe.buffer.as_ptr() as *const c_void;`